File: api/app/models/subject.py
"""
Subject and Chapter models - Maps to Symfony entities
"""
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, UniqueConstraint, JSON
from sqlalchemy.orm import relationship
from datetime import datetime
from app.database import Base


class Subject(Base):
    """Maps to Symfony App.Entity.Subject"""
    __tablename__ = "subjects"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False)
    code = Column(String(50), unique=True, nullable=False, index=True)
    description = Column(Text, nullable=True)
    created_by_id = Column("created_by_id", Integer, ForeignKey("users.id"))
    created_at = Column("created_at", DateTime, default=datetime.utcnow)

    # Relationships
    created_by = relationship("User", back_populates="subjects")
    chapters = relationship("Chapter", back_populates="subject", order_by="Chapter.order_no")
    quizzes = relationship("Quiz", back_populates="subject")
    flashcard_decks = relationship("FlashcardDeck", back_populates="subject")
    revision_plans = relationship("RevisionPlan", back_populates="subject")


class Chapter(Base):
    """Maps to Symfony App.Entity.Chapter"""
    __tablename__ = "chapters"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(255), nullable=False)
    order_no = Column("order_no", Integer, nullable=False, default=1)
    summary = Column(Text, nullable=True)
    content = Column(Text, nullable=True)
    ai_summary = Column("ai_summary", Text, nullable=True)
    ai_key_points = Column("ai_key_points", JSON, nullable=True)
    ai_tags = Column("ai_tags", JSON, nullable=True)
    subject_id = Column(Integer, ForeignKey("subjects.id"), nullable=False)
    created_by_id = Column("created_by_id", Integer, ForeignKey("users.id"))
    created_at = Column("created_at", DateTime, default=datetime.utcnow)

    # Relationships
    subject = relationship("Subject", back_populates="chapters")
    quizzes = relationship("Quiz", back_populates="chapter")
    flashcard_decks = relationship("FlashcardDeck", back_populates="chapter")

    __table_args__ = (
        UniqueConstraint('subject_id', 'order_no', name='unique_subject_order'),
    )

File: src/Entity/Subject.php
<?php

namespace App\Entity;

use App\Repository\SubjectRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: SubjectRepository::class)]
#[ORM\Table(name: 'subjects')]
class Subject
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: Types::INTEGER)]
    private ?int $id = null;

    #[ORM\Column(type: Types::STRING, length: 255)]
    #[Assert\NotBlank]
    #[Assert\Length(max: 120)]
    private string $name;

    #[ORM\Column(type: Types::STRING, length: 50, unique: true, nullable: true)]
    #[Assert\Length(max: 30)]
    private ?string $code = null;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    private ?string $description = null;

    #[ORM\ManyToOne(targetEntity: User::class, inversedBy: 'subjects')]
    #[ORM\JoinColumn(nullable: false)]
    private User $createdBy;

    #[ORM\Column(type: Types::DATETIME_IMMUTABLE)]
    private \DateTimeImmutable $createdAt;

    /** @var Collection<int, Chapter> */
    #[ORM\OneToMany(targetEntity: Chapter::class, mappedBy: 'subject', cascade: ['persist', 'remove'])]
    private Collection $chapters;

    /** @var Collection<int, Quiz> */
    #[ORM\OneToMany(targetEntity: Quiz::class, mappedBy: 'subject')]
    private Collection $quizzes;

    /** @var Collection<int, FlashcardDeck> */
    #[ORM\OneToMany(targetEntity: FlashcardDeck::class, mappedBy: 'subject')]
    private Collection $flashcardDecks;

    /** @var Collection<int, RevisionPlan> */
    #[ORM\OneToMany(targetEntity: RevisionPlan::class, mappedBy: 'subject')]
    private Collection $revisionPlans;

    public function __construct()
    {
        $this->createdAt = new \DateTimeImmutable();
        $this->chapters = new ArrayCollection();
        $this->quizzes = new ArrayCollection();
        $this->flashcardDecks = new ArrayCollection();
        $this->revisionPlans = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function setName(string $name): static
    {
        $this->name = $name;
        return $this;
    }

    public function getCode(): ?string
    {
        return $this->code;
    }

    public function setCode(?string $code): static
    {
        $this->code = $code;
        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(?string $description): static
    {
        $this->description = $description;
        return $this;
    }

    public function getCreatedBy(): User
    {
        return $this->createdBy;
    }

    public function setCreatedBy(User $createdBy): static
    {
        $this->createdBy = $createdBy;
        return $this;
    }

    public function getCreatedAt(): \DateTimeImmutable
    {
        return $this->createdAt;
    }

    /** @return Collection<int, Chapter> */
    public function getChapters(): Collection
    {
        return $this->chapters;
    }

    public function addChapter(Chapter $chapter): static
    {
        if (!$this->chapters->contains($chapter)) {
            $this->chapters->add($chapter);
            $chapter->setSubject($this);
        }
        return $this;
    }

    public function removeChapter(Chapter $chapter): static
    {
        $this->chapters->removeElement($chapter);
        return $this;
    }

    /** @return Collection<int, Quiz> */
    public function getQuizzes(): Collection
    {
        return $this->quizzes;
    }

    /** @return Collection<int, FlashcardDeck> */
    public function getFlashcardDecks(): Collection
    {
        return $this->flashcardDecks;
    }

    /** @return Collection<int, RevisionPlan> */
    public function getRevisionPlans(): Collection
    {
        return $this->revisionPlans;
    }
}
